src/torchcodec/samplers/_implem.py

```diff
@@ -323,3 +331,236 @@ def clips_at_regular_indices(
+            start_seconds + i * seconds_between_frames
+            for i in range(num_frames_per_clip)
+        ]
+        clip_timestamps = [
+            timestamp for timestamp in clip_timestamps if timestamp < end_stream_seconds
+        ]
+
+        if len(clip_timestamps) < num_frames_per_clip:
+            clip_timestamps = policy_fun(clip_timestamps, num_frames_per_clip)
+        all_clips_timestamps += clip_timestamps
+
+    return all_clips_timestamps
+
+
+def _decode_all_clips_timestamps(
+    decoder: VideoDecoder, all_clips_timestamps: list[float], num_frames_per_clip: int
+) -> list[FrameBatch]:
+    # This is 99% the same as _decode_all_clips_indices. The only change is the
+    # call to .get_frame_displayed_at(pts) instead of .get_frame_at(idx)
+
+    def chunk_list(lst, chunk_size):
+        # return list of sublists of length chunk_size
+        return [lst[i : i + chunk_size] for i in range(0, len(lst), chunk_size)]
+
+    def to_framebatch(frames: list[Frame]) -> FrameBatch:
+        # IMPORTANT: see other IMPORTANT note below
+        data = torch.stack([frame.data for frame in frames])
+        pts_seconds = torch.tensor([frame.pts_seconds for frame in frames])
+        duration_seconds = torch.tensor([frame.duration_seconds for frame in frames])
+        return FrameBatch(
+            data=data, pts_seconds=pts_seconds, duration_seconds=duration_seconds
+        )
+
+    all_clips_timestamps_sorted, argsort = zip(
+        *sorted(
+            (frame_index, i) for (i, frame_index) in enumerate(all_clips_timestamps)
+        )
+    )
+    previous_decoded_frame = None
+    all_decoded_frames = [None] * len(all_clips_timestamps)
+    for i, j in enumerate(argsort):
+        frame_pts_seconds = all_clips_timestamps_sorted[i]
+        if (
+            previous_decoded_frame is not None  # then we know i > 0
+            and frame_pts_seconds == all_clips_timestamps_sorted[i - 1]
+        ):
+            # Avoid decoding the same frame twice.
+            # IMPORTANT: this is only correct because a copy of the frame will
+            # happen within `to_framebatch` when we call torch.stack.
+            # If a copy isn't made, the same underlying memory will be used for
+            # the 2 consecutive frames. When we re-write this, we should make
+            # sure to explicitly copy the data.
+            decoded_frame = previous_decoded_frame
+        else:
+            decoded_frame = decoder.get_frame_displayed_at(seconds=frame_pts_seconds)
+        previous_decoded_frame = decoded_frame
+        all_decoded_frames[j] = decoded_frame
+
+    all_clips: list[list[Frame]] = chunk_list(
+        all_decoded_frames, chunk_size=num_frames_per_clip
+    )
+
+    return [to_framebatch(clip) for clip in all_clips]
+
+
+def clips_at_regular_timestamps(
+    decoder,
+    *,
+    seconds_between_clip_starts: float,
+    num_frames_per_clip: int = 1,
+    seconds_between_frames: Optional[float] = None,
```

Comment:
This is a bit misleading because the seconds between frames within a clip are different from across clips.

Maybe rename this to seconds_between_frames_within_clip

Reply:
The same applies to the existing `num_indices_between_frames` parameter that exists for index-based samplers. I don't think it's worth extending this already-long parameter name, it should be obvious from the docs what it means.